### PR TITLE
HDDS-11548. Add some logging to the StateMachine

### PR DESCRIPTION
**Pull request by jianghuazhu**
## What changes were proposed in this pull request?
The purpose of this PR is to add some logs to some StateMachines.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-11548

## How was this patch tested?
https://github.com/jianghuazhu/ozone/actions/runs/11255835628

### CODE REVIEW

**Comment by jianghuazhu** [comment]
@szetszwo  @adoroszlai , can you help review this PR?
Thanks.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java [hunk 1]
```diff
@@ -136,6 +137,8 @@ public void initialize(RaftServer server, RaftGroupId id,
       super.initialize(server, id, raftStorage);
       this.raftGroupId = id;
       storage.init(raftStorage);
+      LOG.info("OzoneManagerStateMachine is initializing. raftPeerId = {}, raftGroupId = {}",
+          server.getId(), raftGroupId);
```
Comment (szetszwo) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java:141]: - Print peer id first.  Let's do it for all the messages, it will be easier for debugging test output since there are more than one peers.
- Could you also remove `raftGroupId` and replace the usage with `getGroupId()`?
```java
@@ -88,7 +89,6 @@ public class OzoneManagerStateMachine extends BaseStateMachine {
       new SimpleStateMachineStorage();
   private final OzoneManager ozoneManager;
   private RequestHandler handler;
-  private RaftGroupId raftGroupId;
   private volatile OzoneManagerDoubleBuffer ozoneManagerDoubleBuffer;
   private final ExecutorService executorService;
   private final ExecutorService installSnapshotExecutor;
@@ -134,8 +134,8 @@ public void initialize(RaftServer server, RaftGroupId id,
       RaftStorage raftStorage) throws IOException {
     getLifeCycle().startAndTransition(() -> {
       super.initialize(server, id, raftStorage);
-      this.raftGroupId = id;
       storage.init(raftStorage);
+      LOG.info("{}: initialize {} with {}", id, getGroupId(), getLastAppliedTermIndex());
     });
   } 
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java [hunk 2]
```diff
@@ -145,6 +148,8 @@ public synchronized void reinitialize() throws IOException {
     if (getLifeCycleState() == LifeCycle.State.PAUSED) {
       unpause(getLastAppliedTermIndex().getIndex(),
           getLastAppliedTermIndex().getTerm());
+      LOG.info("OzoneManagerStateMachine is reinitializing. raftPeerId = {}, raftGroupId = {}, " +
+          "lastAppliedTermIndex = {}", getId(), getGroupId(), getLastAppliedTermIndex());
```
Comment (szetszwo) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java:152]: It should not call `getLastAppliedTermIndex()` more than once.
```java
@@ -143,8 +143,9 @@ public void initialize(RaftServer server, RaftGroupId id,
   public synchronized void reinitialize() throws IOException {
     loadSnapshotInfoFromDB();
     if (getLifeCycleState() == LifeCycle.State.PAUSED) {
-      unpause(getLastAppliedTermIndex().getIndex(),
-          getLastAppliedTermIndex().getTerm());
+      final TermIndex lastApplied = getLastAppliedTermIndex();
+      unpause(lastApplied.getIndex(), lastApplied.getTerm());
+      LOG.info("{}: reinitialize {} with {}", getId(), getGroupId(), lastApplied);
     }
   }
 ```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java [hunk 3]
```diff
@@ -160,6 +165,8 @@ public void notifyLeaderChanged(RaftGroupMemberId groupMemberId,
                                   RaftPeerId newLeaderId) {
     // Initialize OMHAMetrics
     ozoneManager.omHAMetricsInit(newLeaderId.toString());
+    LOG.info("New Leader changes. RaftGroupMemberId = {}, newLeader = {}",
+        groupMemberId, newLeaderId);
```
Comment (szetszwo) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java:169]: Use a shorter message:
```java
    LOG.info("{}: leader changed to {}", groupMemberId, newLeaderId);
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java [hunk 4]
```diff
@@ -293,6 +306,12 @@ public TransactionContext preAppendTransaction(TransactionContext trx)
 
     OzoneManagerPrepareState prepareState = ozoneManager.getPrepareState();
 
+    if (LOG.isDebugEnabled()) {
+      long term = trx.getLogEntry() != null ? trx.getLogEntry().getTerm() : -1;
+      long index = trx.getLogEntry() != null ? trx.getLogEntry().getIndex() : -1;
+      LOG.debug("Preprocess a new transaction. term = {}, index = {}", term, index);
+    }
```
Comment (szetszwo) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java:313]: Use `TermIndex.valueOf`:
```java
     if (LOG.isDebugEnabled()) {
      LOG.debug("{}: preAppendTransaction {}", getId(), TermIndex.valueOf(trx.getLogEntry()));
    }
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java [hunk 5]
```diff
@@ -276,13 +283,19 @@ public TransactionContext startTransaction(
...
         .setStateMachine(this)
         .setServerRole(RaftProtos.RaftPeerRole.LEADER)
         .setLogData(raftClientRequest.getMessage().getContent())
         .setStateMachineContext(omRequest)
         .build();
+    if (LOG.isDebugEnabled()) {
+      long term = ctxt.getLogEntry() != null ? ctxt.getLogEntry().getTerm() : -1;
+      long index = ctxt.getLogEntry() != null ? ctxt.getLogEntry().getIndex() : -1;
+      LOG.debug("Start a new transaction. term = {}, index = {}", term, index);
+    }
```
Comment (szetszwo) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java:297]: Revert this change since `startTransaction` neither has log entry nor term-index.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java [hunk 6]
```diff
@@ -303,8 +322,7 @@ public TransactionContext preAppendTransaction(TransactionContext trx)
       if (ozoneManager.getAclsEnabled()
           && !ozoneManager.isAdmin(userGroupInformation)) {
         String message = "Access denied for user " + userGroupInformation
-            + ". "
-            + "Superuser privilege is required to prepare ozone managers.";
+            + ". Superuser privilege is required to prepare ozone managers.";
```
Comment (szetszwo) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java:325]: Let's change the message to 
```java
            + ". Superuser privilege is required to prepare upgrade/downgrade.";
```

**Review by szetszwo — COMMENTED** [review]
@jianghuazhu , thanks for working on this!

Please see the comments inlined.  In general, let's print an id in the beginning of a message and use shorter message.

**Comment by jianghuazhu** [comment]
Thanks @szetszwo  for the comment and review.
I have updated it.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMHADBTransactionBufferImpl.java [hunk 7]
```diff
@@ -107,6 +111,7 @@ public SnapshotInfo getLatestSnapshot() {
 
   @Override
   public void setLatestSnapshot(SnapshotInfo latestSnapshot) {
+    LOG.info("Set latest Snapshot to {}", latestSnapshot);
```
Comment (szetszwo) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMHADBTransactionBufferImpl.java:114]: Please add server id.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java [hunk 8]
```diff
@@ -149,6 +150,9 @@ public CompletableFuture<Message> applyTransaction(
       final SCMRatisRequest request = SCMRatisRequest.decode(
           Message.valueOf(trx.getStateMachineLogEntry().getLogData()));
 
+      if (LOG.isDebugEnabled()) {
+        LOG.debug("applyTransaction {}", TermIndex.valueOf(trx.getLogEntry()));
```
Comment (szetszwo) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:154]: Please add server id.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java [hunk 9]
```diff
@@ -341,6 +346,7 @@ public CompletableFuture<Message> applyTransaction(TransactionContext trx) {
           : OMRatisHelper.convertByteStringToOMRequest(
           trx.getStateMachineLogEntry().getLogData());
       final TermIndex termIndex = TermIndex.valueOf(trx.getLogEntry());
+      LOG.debug("applyTransaction {}", termIndex);
```
Comment (szetszwo) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java:349]: Please add server id.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java [hunk 10]
```diff
@@ -427,12 +433,14 @@ public synchronized void pause() {
    */
   public synchronized void unpause(long newLastAppliedSnaphsotIndex,
       long newLastAppliedSnapShotTermIndex) {
-    LOG.info("OzoneManagerStateMachine is un-pausing");
     if (statePausedCount.decrementAndGet() == 0) {
       getLifeCycle().startAndTransition(() -> {
         this.ozoneManagerDoubleBuffer = buildDoubleBufferForRatis();
         this.setLastAppliedTermIndex(TermIndex.valueOf(
             newLastAppliedSnapShotTermIndex, newLastAppliedSnaphsotIndex));
+        LOG.info("OzoneManagerStateMachine un-pause completed. " +
```
Comment (szetszwo) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java:441]: Please add server id.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java [hunk 11]
```diff
@@ -482,15 +490,15 @@ private synchronized long takeSnapshotImpl() throws IOException {
...
-    LOG.info(" skipped = {}", lastSkippedIndex);
-    LOG.info("notified = {}", notified);
-    LOG.info("snapshot = {}", snapshot);
 
+    long startTime = Time.monotonicNow();
     final TransactionInfo transactionInfo = TransactionInfo.valueOf(snapshot);
     ozoneManager.setTransactionInfo(transactionInfo);
     ozoneManager.getMetadataManager().getTransactionInfoTable().put(TRANSACTION_INFO_KEY, transactionInfo);
     ozoneManager.getMetadataManager().getStore().flushDB();
+    LOG.info("Taking snapshot. applied = {}, skipped = {}, " +
```
Comment (szetszwo) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java:499]: Please add server id.

**Review by szetszwo — COMMENTED** [review]
@jianghuazhu , thanks for the update!  The change looks good.  Just found that the server id was missing in a few places.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.
